components/message_identifier_router.jsx

```diff
@@ -0,0 +1,218 @@
+        }
+        user = data;
+    }
+
+    openDirectChannelToUser(
+        user.id,
+        (channel) => {
+            doChannelChange(channel);
+        },
+        () => handleError(match, history)
+    );
+}
+
+async function goToDirectChannelByUserId(match, history) {
+    const {team, identifier} = match.params;
+    const userId = identifier.toLowerCase();
+
+    let user = UserStore.getProfile(userId);
+    if (!user) {
+        const {data, error} = await getUser(userId)(dispatch, getState);
+        if (error) {
+            handleError(match, history);
+            return;
+        }
+        user = data;
+    }
+
+    history.replace(`/${team}/messages/@${user.username}`);
+}
+
+async function goToDirectChannelByUserIds(match, history) {
+    const {team, identifier} = match.params;
+    const userId = Utils.getUserIdFromChannelId(identifier.toLowerCase());
+
+    let user = UserStore.getProfile(userId);
+    if (!user) {
+        const {data, error} = await getUser(userId)(dispatch, getState);
+        if (error) {
+            handleError(match, history);
+            return;
+        }
+        user = data;
+    }
+
+    history.replace(`/${team}/messages/@${user.username}`);
+}
+
+async function goToDirectChannelByEmail(match, history) {
+    const {team, identifier} = match.params;
+    const email = identifier.toLowerCase();
+
+    let user = UserStore.getProfileByEmail(email);
+    if (!user) {
+        const {data, error} = await getUserByEmail(email)(dispatch, getState);
+        if (error) {
+            handleError(match, history);
+            return;
+        }
+        user = data;
+    }
+
+    history.replace(`/${team}/messages/@${user.username}`);
+}
+
+async function goToGroupChannelByGroupId(match, history) {
+    const {identifier} = match.params;
+    const groupId = identifier.toLowerCase();
+
+    history.replace(match.url.replace('/channels/', '/messages/'));
+
+    let channel = ChannelStore.getByName(groupId);
+    if (!channel) {
+        const {data, error} = await joinChannel(UserStore.getCurrentId(), TeamStore.getCurrentId(), null, groupId)(dispatch, getState);
+        if (error) {
+            handleError(match, history);
+            return;
+        }
+        channel = data.channel;
+    }
+
+    doChannelChange(channel);
+}
+
+function doChannelChange(channel) {
+    GlobalActions.emitChannelClickEvent(channel);
+}
+
+function handleError(match, history) {
+    const {team} = match.params;
+    history.push(team ? `/${team}/channels/${Constants.DEFAULT_CHANNEL}` : '/');
+}
+
+export default class NeedsTeam extends React.PureComponent {
+    constructor(props) {
+        super(props);
+
+        onChannelByIdentifierEnter(props);
+    }
+
+    componentWillReceiveProps(nextProps) {
+        if (this.props.match.params.identifier !== nextProps.match.params.identifier) {
+            onChannelByIdentifierEnter(nextProps);
+        }
+    }
+
+    render() {
+        return <ChannelView/>;
```

Comment:
Is there a case where we actually render somthing or do we just redirect? If we just redirect then maybe we should just render nil here.

Reply:
There is no redirect here, so I think it should keep returning channel view.

Actually, I'm not sure it can be called "router".

It is just a bunch of onEnter callback to select a channel and it does call an action, does not switch a route.